From 328a4ddbd25dd16c6735a1733b4fcdb0bc21591c Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Mon, 31 May 2021 17:17:15 +0300
Subject: [PATCH] [SW-44018] gaudi: moving dev_id fuse read code for reuse

Both the present preboot_rate_limiter_init and the future to come
preboot_hbm_credits_init functions are using device id data from
fuse for internal logic.
Moving the code retrieving this data to a separate function
for future code reuse.

Change-Id: I124096537da9e9c524cd1fe107029aa15ab6d240
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/gaudi/asic.c b/gaudi/asic.c
index a6d6aca..5c51dca 100644
--- a/gaudi/asic.c
+++ b/gaudi/asic.c
@@ -455,6 +455,26 @@
 	return true;
 }
 
+#define GAUDI_FUSE_DEV_ID_WORD_OFFSET	2
+
+static int hl_fuse_get_dev_id(char *dev_maj, char *dev_min)
+{
+	int ret;
+	u32 dev_type;
+
+	ret = preboot_fuse_read(GAUDI_FUSE_DEV_ID_WORD_OFFSET, &dev_type);
+	if (ret)
+		return ret;
+
+	dev_type = __builtin_bswap32(dev_type);
+	*dev_maj = (u8)((dev_type >> 8) & 0xff);
+	*dev_min = (u8)(dev_type & 0xff);
+	if (!*dev_min)
+		*dev_min = '0';
+
+	return ret;
+}
+
 /*
  * rate limiter parameters for 800MHz clock. that's the only clock currently
  * supported for gaudi hbm in preboot
@@ -466,23 +486,16 @@
 #define GAUDI_HBM_MID_WGHT		0x111
 #define GAUDI_HBM_CH_DIST		0x20000
 #define GAUDI_HBM_NUM_DEVICES		4
-#define GAUDI_FUSE_DEV_ID_WORD_OFFSET	2
 
 void preboot_rate_limiter_init(void)
 {
 	int i, ret;
-	u32 base, dev_type;
+	u32 base;
 	char dev_maj, dev_min;
 
-	ret = preboot_fuse_read(GAUDI_FUSE_DEV_ID_WORD_OFFSET, &dev_type);
+	ret = hl_fuse_get_dev_id(&dev_maj, &dev_min);
 	if (ret)
 		return;
-
-	dev_type = __builtin_bswap32(dev_type);
-	dev_maj = (u8)((dev_type >> 8) & 0xff);
-	dev_min = (u8)(dev_type & 0xff);
-	if (!dev_min)
-		dev_min = '0';
 
 	/* rate limiters are initialized for A0 and B0 devices only */
 	if ((dev_min != '0') || ((dev_maj != 'A') && (dev_maj != 'B')))
